File: Lexer/Cpp/lexer.cpp
#include "lexer.hpp"

Lexer::Lexer(string input) {
    this->input = input;
    this->curr  = 0;
    this->peek  = 0;
    this->ch    = '\0';

    this->readChar();
}

Token Lexer::nextToken() {
    while (isspace(this->ch))
        this->readChar();

    Token tok = Token(::_EOF, "\0");
    switch (this->ch) {
        case '\0': tok = Token(::_EOF, "\0"); break;
        case '*':  tok = Token(::ASTERISK, "*"); break;
        case '/':  tok = Token(::SLASH, "/"); break;
        case '-':  tok = Token(::DASH, "-"); break;
        case '+':  tok = Token(::PLUS, "+"); break;
        case '=':  tok = Token(::EQUAL, "="); break;
        case '<':  tok = Token(::LESSTHAN, "<"); break;
        case '>':  tok = Token(::GREATERTHAN, ">"); break;
        case ';':  tok = Token(::SEMICOLON, ";"); break;
        case ':':  tok = Token(::COLON, ":"); break;
        case '(':  tok = Token(::LPAREN, "("); break;
        case ')':  tok = Token(::RPAREN, ")"); break;
        case '{':  tok = Token(::LBRACE, "{"); break;
        case '}':  tok = Token(::RBRACE, "}"); break;
        default:
            if (isalpha(this->ch)) {
                string ident = this->readIdentifier();
                if (IDENTIFIER_MAP.find(ident) == IDENTIFIER_MAP.end())
                    return Token(::IDENTIFIER, ident);
                else return Token(IDENTIFIER_MAP.at(ident), ident);
            } else if (isdigit(this->ch)) return Token(::NUMBER, this->readNumber());
            else tok = Token(::ILLEGAL, "ILLEGAL");
    };
    this->readChar();
    return tok;
}

string Lexer::readIdentifier() {
    uint64_t pos = this->curr;
    while (isalpha(this->ch) || this->ch == '_')
        this->readChar();
    return this->input.substr(pos, this->curr - pos);
}

string Lexer::readNumber() {
    uint64_t pos = this->curr;
    while (isdigit(this->ch))
        this->readChar();
    return this->input.substr(pos, this->curr - pos);
}

void Lexer::readChar() {
    this->ch   = this->peek >= this->input.length() ? 0 : this->input[this->peek];
    this->curr = this->peek++;
}

File: Lexer/Python/lexer.py
from tokens import IDENTIFIER_MAP, Token, TokenType


class Lexer:
    curr: int
    peek: int
    input: str
    ch: str | None

    def __init__(self, input: str) -> None:
        self.input = input
        self.curr = 0
        self.peek = 0
        self.ch = None
        self.read_char()

    def next_token(self) -> Token:
        tok: Token = Token(TokenType.EOF, "")

        while self.ch is not None and self.ch.isspace():
            self.read_char()

        if not self.ch:
            return tok

        match self.ch:
            case ";":
                tok = Token(TokenType.SEMICOLON, ";")
            case "*":
                tok = Token(TokenType.ASTERISK, "*")
            case "/":
                tok = Token(TokenType.SLASH, "/")
            case "-":
                tok = Token(TokenType.DASH, "-")
            case "+":
                tok = Token(TokenType.PLUS, "+")
            case "=":
                tok = Token(TokenType.EQUAL, "=")
            case "<":
                tok = Token(TokenType.LESSTHAN, "<")
            case ">":
                tok = Token(TokenType.GREATERTHAN, ">")
            case ";":
                tok = Token(TokenType.SEMICOLON, ";")
            case ":":
                tok = Token(TokenType.COLON, ":")
            case "(":
                tok = Token(TokenType.LPAREN, "(")
            case ")":
                tok = Token(TokenType.RPAREN, ")")
            case "{":
                tok = Token(TokenType.LBRACE, "{")
            case "}":
                tok = Token(TokenType.RBRACE, "}")
            case _:
                if self.ch.isalpha():
                    if ttype := IDENTIFIER_MAP.get((ident := self.read_identifier())):
                        return Token(ttype, ident)
                    else:
                        return Token(TokenType.IDENTIFIER, ident)
                elif self.ch.isdigit():
                    return Token(TokenType.NUMBER, self.read_number())
                return Token(TokenType.ILLEGAL, "ILLEGAL")

        self.read_char()
        return tok

    def read_identifier(self) -> str:
        pos = self.curr
        while self.ch is not None and self.ch.isalpha() or self.ch == "_":
            self.read_char()
        return self.input[pos : self.curr]

    def read_number(self) -> str:
        pos = self.curr
        while self.ch is not None and self.ch.isdigit():
            self.read_char()
        return self.input[pos : self.curr]

    def read_char(self) -> None:
        if self.peek >= len(self.input):
            self.ch = None
        else:
            self.ch = self.input[self.peek]
        self.curr = self.peek
        self.peek += 1
